openupgradelib/openupgrade_120.py

```diff
@@ -0,0 +1,208 @@
+       class_rm="checkbox radio", class_add="form-check"),
+    _r(selector="div.form-check label", class_add="form-check-label"),
+    _r(selector="div.form-check input", class_add="form-check-input"),
+    _r(selector="label.checkbox-inline",
+       class_rm="checkbox-inline",
+       class_add="form-check-label form-check-inline"),
+    _r(selector=".checkbox-inline input",
+       class_rm="radio checkbox", class_add="form-check-input"),
+    _r(selector=".form-horizontal .form-group", class_add="row"),
+    _r(selector=".form-horizontal .form-group .control-label",
+       class_rm="control-label", class_add="col-form-label text-right"),
+    _r(selector=".form-horizontal", class_rm="form-horizontal"),
+    _r("form-control-static", "form-control-plaintext"),
+
+    # Input groups
+    _r(selector=".form-control + .input-group-addon",
+       class_rm="input-group-addon", class_add="input-group-text",
+       wrap="<span class='input-group-append'/>"),
+    _r(selector=".form-control + .input-group-btn",
+       class_rm="input-group-btn", class_add="input-group-append"),
+    _r(selector=".input-group-addon",
+       class_rm="input-group-addon", class_add="input-group-text",
+       wrap="<span class='input-group-prepend'/>"),
+    _r(selector=".input-group-btn",
+       class_rm="input-group-btn", class_add="input-group-prepend"),
+
+    # Buttons
+    _r("btn-default", "btn-secondary"),
+    _r("btn-xs", "btn-sm"),
+    _r("btn-group-xs", "btn-group-sm"),
+    _r("btn-group-justified", "w-100",
+       wrap='<div class="btn-group d-flex" role="group"/>'),
+    _r(selector=".btn-group + .btn-group", class_add="ml-1"),
+
+    # Dropdowns
+    _r("divider", "dropdown-divider", selector=".dropdown-menu > .divider"),
+    _r(selector=".dropdown-menu > li > a", class_add="dropdown-item"),
+
+    # List groups
+    _r("list-group-item", "list-group-item-action",
+       selector="a.list-group-item"),
+
+    # Navs
+    _r(selector=".nav > li", class_add="nav-item"),
+    _r(selector=".nav > li > a", class_add="nav-link"),
+    _r("nav-stacked", "flex-column"),
+
+    # Navbar
+    _r(selector="navbar", class_add="navbar-expand-sm"),
+    _r("navbar-default", "navbar-light"),
+    _r("navbar-toggle", "navbar-toggler"),
+    _r("navbar-form", "form-inline"),
+    _r("navbar-fixed-top", "fixed-top"),
+    _r("navbar-btn", "nav-item"),
+    _r("navbar-right", "ml-auto"),
+
+    # Pagination
+    _r(selector=".pagination > li", class_add="page-item"),
+    _r(selector=".pagination > li > a", class_add="page-link"),
+
+    # Breadcrumbs
+    _r(selector=".breadcrumb > li", class_add="breadcrumb-item"),
+
+    # Labels and badges
+    _r("label", "badge"),
+    _r("badge-default", "badge-secondary"),
+    *(_r("label-%s" % c3, "badge-%s" % c4) for (c3, c4) in CONTEXTS),
+
+    # Convert panels, thumbnails and wells to cards
+    _r("panel", "card"),
+    _r("panel-body", "card-body"),
+    _r("panel-default panel-group"),
```

Comment:
You need to put `", "` here.

**Remark:** Also, as an overall comment for all this code, you should also consider what Odoo did when it changed to BS4. I mean, here in openupgradelib we should not have a pure BS3 to BS4 converter, here we should have a Odoo BS3 to Odoo BS4 converter instead.

As an example of what I mean, Odoo didn't do:
```
_r("panel-default", "panel-group"),
```
What Odoo did was:
```
_r("panel-collapse", ""),
_r("panel-default", ""),
_r("panel-group", "accordion"),
```

Reply:
for `_r("panel-default", "")` and `_r("panel-group", "accordion")`, you can find examples here:
https://github.com/odoo/odoo/commit/59237ea81988bdcd0cc746422a7c94bf2eadaaa2#diff-24b0433302053618c5b742f3eded1d67

As for the `panel-collapse`, I found that in general the change is `_r("panel-collapse", "collapse")` but in some cases is `_r("panel-collapse", "")`. The same goes to wells, sometimes  is `_r("well", "card card-body")`, but sometimes is  `_r("well", "")`.

Reply:
I took that into account, although the goal here is not mainly to format widgets as the new widgets, but to format them as close as possible to how they were before.

If a user wants to update the look and feel of his website, usually he should drag and drop a new snippet and modify its content, and remove the old one.

Thanks!

Reply:
@Yajo but you should have something into account: old snippets must be editable the same with new version, so if a snippet in v11 was for example image gallery, on v12 it should still be recognized by web_editor as image gallery. Take that into account.

Reply:
@pedrobaeza exactly, that was the intention of my remark, good explanation 😉

@Yajo still you should add, at least, the `", "` in that indicated line :)

Reply:
> old snippets must be editable the same with new version

Yes, they are, I just meant that it's more important to keep the visual changes minimal. Users visit their pages more often than they modify them.

I'll double-check this manually however. Set this as WIP again in the mean time, please.



> you should add, at least, the `", "` in that indicated line :)

Well, with that I'd be replacing the `panel-default` class for a `panel-group`, but quoting from the [BS migration docs](https://getbootstrap.com/docs/4.3/migration/#panels):

> - `.panel-default` removed and no replacement.
> - `.panel-group` removed and no replacement. `.card-group` is not a replacement, it is different.

So, my intention was to just drop both. I have to make a change, but it's different; thanks for the pointer anyways. :blush:

Reply:
Thanks for your comments, there were indeed some usability glitches, they are polished now. :wink:  I'm gonna upload some gifs, down in the main thread.